File: R/spatvec.R



setMethod("geomtype", signature(x="SpatVector"),
	function(x){
		x@ptr$type()
	}
)
setMethod("geomtype", signature(x="SpatVectorProxy"),
	function(x){
		x@ptr$v$type()
	}
)


setMethod("datatype", signature(x="SpatVector"),
	function(x){
		x@ptr$df$get_datatypes()
	}
)


setMethod("is.lines", signature(x="SpatVector"),
	function(x) {
		geomtype(x) == "lines"
	}
)

setMethod("is.polygons", signature(x="SpatVector"),
	function(x) {
		geomtype(x) == "polygons"
	}
)
setMethod("is.points", signature(x="SpatVector"),
	function(x) {
		grepl("points", geomtype(x))
	}
)


setMethod("geomtype", signature(x="Spatial"),
	function(x){
		type <- sub("spatial", "", as.vector(tolower(class(x)[1])))
		type <- sub("dataframe", "", type)
		if (type %in% c("grid", "pixels")) type <- "raster"
		type
	}
)

setMethod("geom", signature(x="SpatVector"),
	function(x, wkt=FALSE, hex=FALSE, df=FALSE, list=FALSE, xnm="x", ynm="y"){
		if (hex) {
			x@ptr$hex()
		} else if (wkt) {
			x@ptr$getGeometryWKT()
			# or via geos with
			# x@ptr$wkt()
		} else if (list) {
			x@ptr$get_geometryList(xnm, ynm)
		} else {
			g <- x@ptr$get_geometry()
			g <- do.call(cbind, g)
			colnames(g) <- c("geom", "part", "x", "y", "hole")[1:ncol(g)]
			if (df) {
				data.frame(g)
			} else {
				g
			}
		}
	}
)

setMethod("crds", signature(x="SpatVector"),
	function(x, df=FALSE, list=FALSE){
		if (list) {
			gt <- geomtype(x) 
			if (gt == "lines") {
				x@ptr$linesNA()
			} else if (gt == "polygons") {
				x@ptr$polygonsList()			
			} else {
				x@ptr$coordinates()
			}
		} else {
			g <- x@ptr$coordinates()
			g <- do.call(cbind, g)
			colnames(g) <- c("x", "y")
			if (df) {
				data.frame(g)
			} else {
				g
			}
		}
	}
)


setMethod("crds", signature(x="SpatRaster"),
	function(x, df=FALSE, na.rm=TRUE, na.all=FALSE){
#		crds( as.points(x, values=(hasValues(x) || (na.rm)), na.rm=na.rm, na.all=na.all), df=df)
		opt <- spatOptions()
		out <- x@ptr$crds(na.rm, na.all, opt)
		messages(x)
		if (df) {
			out <- data.frame(out)
		} else {
			out <- do.call(cbind, out)
		}
		colnames(out) <- c("x", "y")
		out
	}
)


setMethod("dim", signature(x="SpatVector"),
	function(x){
		c(nrow(x), ncol(x))
	}
)

setMethod("dim", signature(x="SpatVectorProxy"),
	function(x){
		c(x@ptr$v$geom_count, x@ptr$v$ncol())
	}
)


as.data.frame.SpatVector <- function(x, row.names=NULL, optional=FALSE, geom=NULL, ...) {
	d <- .getSpatDF(x@ptr$df, ...)
	# fix empty names
	colnames(d)[1:ncol(x)] <- x@ptr$names
	if (!is.null(geom)) {
		geom <- match.arg(toupper(geom), c("WKT", "HEX", "XY"))
		if (geom == "XY") {
			if (!grepl("points", geomtype(x))) {
				error("as.data.frame", 'geom="XY" is only valid for point geometries')
			}
			if (nrow(d) > 0) {
				d <- cbind(d, crds(x))
			} else {
				d <- data.frame(crds(x), ...)
			}
		} else {
			g <- geom(x, wkt=geom=="WKT", hex=geom=="HEX")
			if (nrow(d) > 0) {
				d$geometry <- g
			} else {
				d <- data.frame(geometry=g, stringsAsFactors=FALSE, ...)
			}
		}
	}
	d
}
setMethod("as.data.frame", signature(x="SpatVector"), as.data.frame.SpatVector)


get.data.frame <- function(x) {
	v <- vect()
	v@ptr <- x@ptr$v
	d <- as.data.frame(v)
	d[0,,drop=FALSE]
}


as.list.SpatVector <- function(x, geom=NULL, ...) {
	as.list(as.data.frame(x, geom=geom))
}
setMethod("as.list", signature(x="SpatVector"), as.list.SpatVector)



setMethod ("expanse", "SpatVector",
	function(x, unit="m", transform=TRUE) {
		a <- x@ptr$area(unit, transform, double());
		x <- messages(x, "expanse");
		return(abs(a))
	}
)


setMethod("perim", signature(x="SpatVector"),
	function(x) {
		p <- x@ptr$length();
		x <- messages(x, "perim");
		return(p)
	}
)

setMethod("length", signature(x="SpatVector"),
	function(x) {
		x@ptr$size()
	}
)


setMethod("fillHoles", signature(x="SpatVector"),
	function(x, inverse=FALSE) {
		if (inverse) {
			x@ptr <- x@ptr$get_holes()
		} else {
			x@ptr <- x@ptr$remove_holes()
		}
		messages(x, "fillHoles")
	}
)



#setMethod("eliminate", signature(x="SpatVector"),
#	function(x, y) {
#		x@ptr <- x@ptr$eliminate(y@ptr)
#		messages(x)
#	}
#)



setMethod("centroids", signature(x="SpatVector"),
	function(x, inside=FALSE) {
		if (inside) {
			x@ptr <- x@ptr$point_on_surface(TRUE)
		} else {
			x@ptr <- x@ptr$centroid(TRUE)
		}
		messages(x, "centroids")
	}
)



setMethod("densify", signature(x="SpatVector"),
	function(x, interval, equalize=TRUE, flat=FALSE) {
		x@ptr <- x@ptr$densify(interval, equalize, flat)
		messages(x, "densify")
	}
)

setMethod("normalize.longitude", signature(x="SpatVector"),
	function(x) {
		if (nrow(x) == 0) return(deepcopy(x))
		nc <- ncol(x)
		fname <- "uuu-_123_uqq_-agg_-id123"
		x[[fname]] <- 1:nrow(x)
		d <- values(x)
		out <- x[,fname]
		out@ptr <- out@ptr$normalize_longitude()
		out <- messages(out)
		a <- aggregate(out, fname, count=FALSE)
		if (nc > 0) {
			a <- merge(a, d, by=fname)
		}
		a[[fname]] <- NULL
		a
	}
)

File: python/src/terra/spatvec.py
"""
spatvec.py — SpatVector geometry access and spatial measurements.
"""
from __future__ import annotations
from typing import List, Optional, Union
import numpy as np

from ._terra import SpatRaster, SpatVector, SpatOptions
from ._helpers import messages, _getSpatDF


def _opt() -> SpatOptions:
    return SpatOptions()


# ---------------------------------------------------------------------------
# geometry type predicates
# ---------------------------------------------------------------------------

def geomtype(x: SpatVector) -> str:
    """Return the geometry type of *x*: ``'points'``, ``'lines'``, or ``'polygons'``."""
    # Must use C++ ``type()``, not ``geomtype()`` — the latter is patched onto
    # SpatVector by register_methods() and delegates back here (recursion).
    return x.type()


def is_lines(x: SpatVector) -> bool:
    """Return True if *x* is a line geometry."""
    return geomtype(x) == "lines"


def is_polygons(x: SpatVector) -> bool:
    """Return True if *x* is a polygon geometry."""
    return geomtype(x) == "polygons"


def is_points(x: SpatVector) -> bool:
    """Return True if *x* is a point geometry."""
    return "points" in geomtype(x)


# ---------------------------------------------------------------------------
# geometry access
# ---------------------------------------------------------------------------

def geom(
    x: SpatVector,
    wkt: bool = False,
    hex: bool = False,
    wkb: bool = False,
    as_df: bool = False,
    as_list: bool = False,
) -> Union[np.ndarray, List[str], List]:
    """
    Return the geometry of *x*.

    Parameters
    ----------
    x : SpatVector
    wkt : bool
        Return WKT strings.
    hex : bool
        Return hexadecimal WKB strings.
    wkb : bool
        Return raw WKB bytes.
    as_df : bool
        Return as a DataFrame (requires pandas).
    as_list : bool
        Return as a list of coordinate arrays.

    Returns
    -------
    numpy.ndarray, list of str, or list depending on arguments.
    """
    if hex:
        return list(x.hex())
    if wkt:
        return list(x.getGeometryWKT())
    if as_list:
        return x.get_geometryList("x", "y")
    if wkb:
        return x.wkb_raw()
    g = x.get_geometry()
    raw = list(g)
    n_cols = len(raw)
    if n_cols == 0:
        return np.empty((0, 5))
    max_len = max(len(col) for col in raw)
    arr = np.full((max_len, n_cols), float("nan"))
    for i, col in enumerate(raw):
        arr[:len(col), i] = col
    col_names = ["geom", "part", "x", "y", "hole"][:n_cols]
    if as_df:
        try:
            import pandas as pd
            return pd.DataFrame(arr, columns=col_names)
        except ImportError:
            pass
    return arr


def crds(
    x: Union[SpatVector, SpatRaster],
    as_df: bool = False,
    as_list: bool = False,
    na_rm: bool = True,
    na_all: bool = False,
) -> Union[np.ndarray, "pd.DataFrame", List]:
    """
    Return the coordinates of *x*.

    Parameters
    ----------
    x : SpatVector or SpatRaster
    as_df : bool
        Return as a DataFrame.
    as_list : bool
        Return as a list of arrays (SpatVector only).
    na_rm : bool
        Skip NA cells (SpatRaster only).
    na_all : bool
        Treat as NA only when all layers are NA (SpatRaster only).

    Returns
    -------
    numpy.ndarray, shape (n, 2), columns [x, y], or DataFrame/list.
    """
    if isinstance(x, SpatRaster):
        opt = _opt()
        out = x.crds(na_rm, na_all, opt)
        messages(x)
        raw = list(out)
        if len(raw) < 2:
            return np.empty((0, 2))
        arr = np.column_stack([raw[0], raw[1]])
    else:
        if as_list:
            gt = geomtype(x)
            if gt == "lines":
                return x.linesNA()
            elif gt == "polygons":
                return x.polygonsList()
            else:
                return x.coordinates()
        raw = x.coordinates()
        raw_list = list(raw)
        if len(raw_list) < 2:
            return np.empty((0, 2))
        arr = np.column_stack([raw_list[0], raw_list[1]])

    if as_df:
        try:
            import pandas as pd
            return pd.DataFrame(arr, columns=["x", "y"])
        except ImportError:
            pass
    return arr


# ---------------------------------------------------------------------------
# spatial measurements
# ---------------------------------------------------------------------------

def expanse(
    x: SpatVector,
    unit: str = "m",
    transform: bool = True,
) -> np.ndarray:
    """
    Compute the area of each feature.

    Parameters
    ----------
    x : SpatVector
        Must contain polygon geometries.
    unit : str
        ``"m"`` (square metres, default) or ``"km"`` (square kilometres).
    transform : bool
        Transform geographic coordinates to an equal-area projection.

    Returns
    -------
    numpy.ndarray of float (one value per feature).
    """
    a = x.area(unit, transform, [])
    messages(x, "expanse")
    return np.abs(np.array(a, dtype=float))


def perim(x: SpatVector) -> np.ndarray:
    """
    Compute the perimeter (or length) of each feature.

    Parameters
    ----------
    x : SpatVector

    Returns
    -------
    numpy.ndarray of float.
    """
    p = x.length()
    messages(x, "perim")
    return np.array(p, dtype=float)


def nseg(x: SpatVector) -> np.ndarray:
    """Return the number of segments per feature."""
    s = x.nsegments()
    messages(x, "nseg")
    return np.array(s, dtype=int)


# ---------------------------------------------------------------------------
# fill holes
# ---------------------------------------------------------------------------

def fill_holes(x: SpatVector, inverse: bool = False) -> SpatVector:
    """
    Fill interior holes in polygon geometries.

    Parameters
    ----------
    x : SpatVector
    inverse : bool
        If True, return only the holes (i.e. extract holes).

    Returns
    -------
    SpatVector
    """
    if inverse:
        xc = x.get_holes()
    else:
        xc = x.remove_holes()
    return messages(xc, "fillHoles")


# ---------------------------------------------------------------------------
# as_data_frame for SpatVector
# ---------------------------------------------------------------------------

def vect_as_df(
    x: SpatVector,
    geom: Optional[str] = None,
) -> "pd.DataFrame":
    """
    Convert the attribute table of *x* to a pandas DataFrame.

    Parameters
    ----------
    x : SpatVector
    geom : str, optional
        If provided, append geometry: ``"WKT"``, ``"HEX"``, or ``"XY"``
        (points only).

    Returns
    -------
    pandas.DataFrame
    """
    try:
        import pandas as pd
    except ImportError:
        raise ImportError("pandas is required for vect_as_df()")

    from ._helpers import _getSpatDF
    d = _getSpatDF(x.df)
    if d is None:
        d = pd.DataFrame()

    if geom is not None:
        geom = geom.upper()
        if geom == "XY":
            if not is_points(x):
                raise ValueError('geom="XY" is only valid for point geometries')
            xy = crds(x, as_df=True)
            d = pd.concat([d.reset_index(drop=True), xy.reset_index(drop=True)], axis=1)
        elif geom in ("WKT", "HEX"):
            d = d.copy()
            d["geometry"] = geom_as_wkt(x) if geom == "WKT" else list(x.hex())
    return d


def geom_as_wkt(x: SpatVector) -> List[str]:
    """Return WKT representation for each feature."""
    return list(x.getGeometryWKT())
